### PR TITLE
Fix S1854 FP: Throw should connect to outer catch

### LINKED ISSUE #9468: Fix S1854 FP: Throw should connect to outer catch
This is similar to #9466 
```
    public void NestedCatchAndRethrow()
    {
        var value = 100;
        try
        {
            try
            {
                CanThrow();
            }
            catch
            {
                value = 200;    // Noncompliant FP
                throw;
            }
        }
        catch
        {
            Log(value);
        }
    }
```

### PR DESCRIPTION
**Pull request by mary-georgiou-sonarsource**
Fixes #9468

### CODE REVIEW

File: analyzers/src/SonarAnalyzer.CFG/LiveVariableAnalysis/RoslynLiveVariableAnalysis.cs [hunk 1]
```diff
@@ -138,6 +138,10 @@ private void BuildBranches(BasicBlock block)
                 AddPredecessorsOutsideRegion(finallyBlock);
             }
         }
+        if (block.IsEnclosedIn(ControlFlowRegionKind.Catch) && block.Successors.FirstOrDefault(x => x.Semantics == ControlFlowBranchSemantics.Rethrow) is { })
```
Comment (sebastien-marichal) [inline analyzers/src/SonarAnalyzer.CFG/LiveVariableAnalysis/RoslynLiveVariableAnalysis.cs:141]:
```suggestion
        if (block.IsEnclosedIn(ControlFlowRegionKind.Catch) && block.Successors.FirstOrDefault(x => x.Semantics == ControlFlowBranchSemantics.Rethrow) is not null)
```

It's probably an FN for T0007.

**Review by sebastien-marichal — APPROVED** [review]
LGTM!
We discussed it together and it seems good for me.
I only added a comment about `is not null` usage

File: analyzers/src/SonarAnalyzer.CFG/LiveVariableAnalysis/RoslynLiveVariableAnalysis.cs [hunk 2]
```diff
@@ -161,6 +165,17 @@ private void BuildBranchesFinally(BasicBlock source, ControlFlowRegion finallyRe
         }
     }
 
+    private void BuildBranchesNestedCatchRethrow(BasicBlock block)
+    {
+        var tryRegion = block.EnclosingRegion(ControlFlowRegionKind.TryAndCatch).NestedRegion(ControlFlowRegionKind.Try);
+        foreach (var catchBlock in tryRegion.ReachableHandlers()
+                                            .Where(x => x.Kind == ControlFlowRegionKind.Catch && x != block.EnclosingRegion && x.FirstBlockOrdinal > block.Ordinal)
```
Comment (Tim-Pohlmann) [inline analyzers/src/SonarAnalyzer.CFG/LiveVariableAnalysis/RoslynLiveVariableAnalysis.cs:172]:
```suggestion
                                            .Where(x => x.Kind == ControlFlowRegionKind.Catch && x.FirstBlockOrdinal > block.Ordinal)
```
The middle condition is already covered by the last condition.

File: analyzers/src/SonarAnalyzer.CFG/LiveVariableAnalysis/RoslynLiveVariableAnalysis.cs [hunk 3]
```diff
@@ -161,6 +165,17 @@ private void BuildBranchesFinally(BasicBlock source, ControlFlowRegion finallyRe
...
+                                            .SelectMany(x => x.Blocks(Cfg)))
```
Comment (Tim-Pohlmann) [inline analyzers/src/SonarAnalyzer.CFG/LiveVariableAnalysis/RoslynLiveVariableAnalysis.cs:173]: I would separate this into an extra variable. `foreach` with multiple lines is always awkward to read.

File: analyzers/tests/SonarAnalyzer.Test/LiveVariableAnalysis/RoslynLiveVariableAnalysisTest.TryCatchFinally.cs [hunk 4]
```diff
@@ -603,6 +603,123 @@ public void Catch_Loop_Propagates_LiveIn()
...
+        context.Validate("Method(3);", LiveIn("value"));
+        context.ValidateExit();
+    }
+
+    [TestMethod]
+    public void Throw_NestedCatch_OuterCatchRethrows_LiveOutOuterCathc()
+    {
+        var code = """
+            var value = 100;
+            try{
```
Comment (Tim-Pohlmann) [inline analyzers/tests/SonarAnalyzer.Test/LiveVariableAnalysis/RoslynLiveVariableAnalysisTest.TryCatchFinally.cs:687]: formatting

File: analyzers/src/SonarAnalyzer.CFG/LiveVariableAnalysis/RoslynLiveVariableAnalysis.cs (hunk 1 above)
Comment (Tim-Pohlmann) [inline analyzers/src/SonarAnalyzer.CFG/LiveVariableAnalysis/RoslynLiveVariableAnalysis.cs:141]:
```suggestion
        if (block.IsEnclosedIn(ControlFlowRegionKind.Catch) && block.Successors.Any(x => x.Semantics == ControlFlowBranchSemantics.Rethrow))
```

**Review by Tim-Pohlmann — APPROVED** [review]
LGTM only cosmetic suggestions.

File: analyzers/src/SonarAnalyzer.CFG/LiveVariableAnalysis/RoslynLiveVariableAnalysis.cs [hunk 5]
```diff
@@ -138,6 +138,10 @@ private void BuildBranches(BasicBlock block)
                 AddPredecessorsOutsideRegion(finallyBlock);
             }
         }
+        if (block.IsEnclosedIn(ControlFlowRegionKind.Catch) && block.Successors.Any(x => x.Semantics == ControlFlowBranchSemantics.Rethrow))
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CFG/LiveVariableAnalysis/RoslynLiveVariableAnalysis.cs:141]: What about `Throw` semantics? All the UTs have `throw`, but `throw new Excetion("Whaaat?", ex)` should behave the same

File: analyzers/src/SonarAnalyzer.CFG/LiveVariableAnalysis/RoslynLiveVariableAnalysis.cs [hunk 6]
```diff
@@ -161,6 +165,15 @@ private void BuildBranchesFinally(BasicBlock source, ControlFlowRegion finallyRe
         }
     }
 
+    private void BuildBranchesNestedCatchRethrow(BasicBlock block)
+    {
+        var reachableHandlers = block.EnclosingRegion(ControlFlowRegionKind.TryAndCatch).NestedRegion(ControlFlowRegionKind.Try).ReachableHandlers();
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CFG/LiveVariableAnalysis/RoslynLiveVariableAnalysis.cs:170]: Doesn't `ReachableHandlers` also connect back to it's own `Catch` in this case?

Comment (mary-georgiou-sonarsource) [reply analyzers/src/SonarAnalyzer.CFG/LiveVariableAnalysis/RoslynLiveVariableAnalysis.cs:170]: @pavel-mikula-sonarsource  I'm after asking for `x.FirstBlockOrdinal > block.Ordinal)` - all the catch with greater ordinal that the one I'm in.
Isn't that working?
```csharp
var reachableHandlers = block.EnclosingRegion(ControlFlowRegionKind.TryAndCatch).NestedRegion(ControlFlowRegionKind.Try).ReachableHandlers();
foreach (var catchBlock in reachableHandlers.Where(x => x.Kind == ControlFlowRegionKind.Catch && x.FirstBlockOrdinal > block.Ordinal).SelectMany(x => x.Blocks(Cfg)))
{}
```

Also not sure what you mean neighbouring catches :/

File: analyzers/src/SonarAnalyzer.CFG/LiveVariableAnalysis/RoslynLiveVariableAnalysis.cs (hunk 5 above)
Comment (mary-georgiou-sonarsource) [reply analyzers/src/SonarAnalyzer.CFG/LiveVariableAnalysis/RoslynLiveVariableAnalysis.cs:141]: fixed here https://github.com/SonarSource/sonar-dotnet/pull/9530
Thanks @sebastien-marichal

File: analyzers/src/SonarAnalyzer.CFG/LiveVariableAnalysis/RoslynLiveVariableAnalysis.cs (hunk 6 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.CFG/LiveVariableAnalysis/RoslynLiveVariableAnalysis.cs:170]:
```VB
Try
Catch Ex As IOException
  Throw 'You are here, with your block.Ordinal
Catch Ex As FormatException
   'This is a neighbour that you don't want to connect to. And it has FirstBlockOrdinal > block.Ordinal
End Try
```

Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.CFG/LiveVariableAnalysis/RoslynLiveVariableAnalysis.cs:170]: Go escape, you'd need to store you're `TryAndCatch` region, and be bigger than its `LastBlockOrdinal`

Comment (mary-georgiou-sonarsource) [reply analyzers/src/SonarAnalyzer.CFG/LiveVariableAnalysis/RoslynLiveVariableAnalysis.cs:170]: Got It! 
I'm opening a PR to fix this.

**Pull request merged**